lib/xandra/protocol.ex

```diff
@@ -831,54 +819,49 @@ defmodule Xandra.Protocol do
     raise "cannot decode custom type #{inspect(class)}"
   end
 
-  defp decode_type_udt(buffer, 0, acc) do
-    {{:udt, Enum.reverse(acc)}, buffer}
+  defp decode_type_udt(<<buffer::bits>>, 0, result) do
+    {{:udt, Enum.reverse(result)}, buffer}
   end
 
-  defp decode_type_udt(buffer, count, acc) do
-    {field_name, buffer} = decode_string(buffer)
+  defp decode_type_udt(<<buffer::bits>>, count, result) do
+    decode_string(field_name <- buffer)
     {field_type, buffer} = decode_type(buffer)
-
-    decode_type_udt(buffer, count - 1, [{field_name, field_type} | acc])
+    decode_type_udt(buffer, count - 1, [{field_name, field_type} | result])
   end
 
-  defp decode_type_tuple(buffer, 0, acc) do
-    {{:tuple, Enum.reverse(acc)}, buffer}
+  defp decode_type_tuple(<<buffer::bits>>, 0, result) do
+    {{:tuple, Enum.reverse(result)}, buffer}
   end
 
-  defp decode_type_tuple(buffer, count, acc) do
+  defp decode_type_tuple(<<buffer::bits>>, count, result) do
     {type, buffer} = decode_type(buffer)
-    decode_type_tuple(buffer, count - 1, [type | acc])
+    decode_type_tuple(buffer, count - 1, [type | result])
   end
 
-  defp decode_string_multimap(<<size::16>> <> buffer) do
-    decode_string_multimap(buffer, size, [])
+  defp decode_string_multimap(<<count::16, buffer::bits>>) do
+    decode_string_multimap(buffer, count, [])
   end
 
-  defp decode_string_multimap(buffer, 0, acc) do
-    {Map.new(acc), buffer}
+  defp decode_string_multimap(<<buffer::bits>>, 0, result) do
+    {Map.new(result), buffer}
   end
 
-  defp decode_string_multimap(buffer, size, acc) do
-    {key, buffer} = decode_string(buffer)
+  defp decode_string_multimap(<<buffer::bits>>, count, result) do
+    decode_string(key <- buffer)
     {value, buffer} = decode_string_list(buffer)
-    decode_string_multimap(buffer, size - 1, [{key, value} | acc])
-  end
-
-  defp decode_string(<<size::16, content::size(size)-bytes>> <> buffer) do
-    {content, buffer}
+    decode_string_multimap(buffer, count - 1, [{key, value} | result])
   end
 
-  defp decode_string_list(<<size::16>> <> buffer) do
-    decode_string_list(buffer, size, [])
+  defp decode_string_list(<<count::16, buffer::bits>>) do
+    decode_string_list(buffer, count, [])
   end
 
-  defp decode_string_list(buffer, 0, acc) do
-    {Enum.reverse(acc), buffer}
+  defp decode_string_list(<<buffer::bits>>, 0, result) do
+    {Enum.reverse(result), buffer}
   end
 
-  defp decode_string_list(buffer, size, acc) do
-    {elem, buffer} = decode_string(buffer)
-    decode_string_list(buffer, size - 1, [elem | acc])
+  defp decode_string_list(<<buffer::bits>>, count, result) do
+    decode_string(item <- buffer)
+    decode_string_list(buffer, count - 1, [item | result])
```

Comment:
Yeah, that `acc` –> `result` change is here for a discussion. I started to use `result` because `acc` feels even more generic, also it is always an issue when you have more than one accumulator. In this case `result` means the result of list decoding, seems better to me than just some accumulator.